### PR TITLE
[ONNX] Handle onnx::Size in ComputeConstant folding

### PR DESCRIPTION
**Pull request by jiafatom**
Support export aten::dim const folding in torchvision.ops.clip_boxes_to_image.
Our traditional const folding cannot handle onnx::Size, because the case here, the input value is unknown. But the new const folding technique can do that, because shape is constant which is enough.

### CODE REVIEW

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 1]
```diff
@@ -6231,6 +6231,21 @@ def forward(self, input):
         self.run_test(DimModel(), empty_input)
         self.run_test(DimModel(), multi_dim_input)
 
+    @skipIfUnsupportedMinOpsetVersion(11)
+    def test_dim_1(self):
+        class M(torch.jit.ScriptModule):
+            @torch.jit.script_method
```
Comment (fatcat-z) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:6237]: Do we need this for forward() method?

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 2]
```diff
@@ -1029,6 +1029,32 @@ void ComputeConstant(Node* n, int opset_version) {
...
+                .sizes();
+        if (input0_shape_size.has_value()) {
+          auto input0_shape_value = input0_shape_size.value();
+          int64_t total_size = 1;
+          auto is_full_static = true;
+          for (auto i = 0; i < input0_shape_value.size(); i++) {
+            if (input0_shape_value[i].is_static()) {
+              total_size *= input0_shape_value[i].static_size();
+            } else {
+              is_full_static = false;
```
Comment (neginraoof) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:1046]: With "break" here in line 1047, does it continue in the loop anymore? Is "is_full_static = false" needed?

Comment (jiafatom) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:1046]: In line 1050, we only set it to const value map when `is_full_static=true`.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 1 above)
Comment (jiafatom) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:6237]: See https://brsoff.github.io/tutorials/advanced/cpp_export.html
"we can convert it to a ScriptModule by subclassing it from torch.jit.ScriptModule and adding a @torch.jit.script_method annotation to the model’s forward method:"

**Review by neginraoof — APPROVED** [review]
nit: If we want to keep the code consistent, maybe handle each op in a separate function similar to ProcessSliceNode.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 1 above)
Comment (fatcat-z) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:6237]: I checked JIT document here https://pytorch.org/docs/stable/jit.html#migrating-to-pytorch-1-2-recursive-scripting-api, and for forward() method, it should be the default behavior so we don't need it anymore.

Anyway, it doesn't hurt by adding this.

File: docs/source/onnx.rst [hunk 3]
```diff
@@ -872,6 +872,101 @@ The interface for specifying operator definitions is experimental;
 adventurous users should note that the APIs will probably
 change in a future interface.
 
+Autograd Function
```
Comment (shubhambhokare1) [inline docs/source/onnx.rst:875]: If possible could you rebase, I'm not exactly sure which changes are related to this PR

Comment (jiafatom) [reply docs/source/onnx.rst]: Thanks, done

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 1 above)
Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:6242]: `@torch.jit.script_method` on forward is required for classes inherited from `torch.jit.ScriptModule`. That being said, it is indeed deprecated, and the preferred way is to inherit from `torch.nn.Module`, and call `torch.jit.script(M())`.

**Review by BowenBao — APPROVED** [review]

**Review by shubhambhokare1 — APPROVED** [review]
LGTM, thanks
